# Proyecto A: Optimización en la Planeación de Transporte Vehicular Urbana Para LogistiCo

## Integrantes

- Barrera Toro, Javier Steven
- Borbon Holguin, Luis Alfredo
- Rolon Toloza, Julian Santiago

## 1.  Formulación Matemática del Modelo

### 1.1. Definición de Conjuntos

- D: Conjunto de centros de distribución:
$$D = \{D1, D2, D3\}$$

- C: Conjunto de zonas de entrega.
$$C = \{C1, C2, C3\}$$

- N Conjunto de todos los nodos.
$$N = C \cup D$$

- T: Conjunto de vehículos disponibles.
$$ T = \{T1, T2, T3\} $$

- A: Conjunto de rutas posibles entre nodos, definidos por:
$$ A = \{(i, j) \in N \times N | i \neq j \} $$

Además, para facilitar el recorrido sobre los conjuntos definimos el siguientes indices.

- $i, j \in N$: índices para representar los nodos (clientes y centros de distribución).
- $t \in T$: Índice para los vehículos disponibles.

### 1.2. Definición de parametros:

Los parámetros que se definen para el modelo son los siguientes: 

- $q_t$: La capacidad máxima de carga del vehículo $t$.
- $c_i$: La capacidad máxima del almacenamiento en el centro de distribución $i \in D$.
- $d_i$: Demanda del cliente $i \in C$.
- $a_t$: Automia maxima de cada vehiculo

Para las distancias y tiempos entre los nodos se definen los siguientes. 
- $d_{ij}$: Distancia entre los nodos $i$ y $j$.
- $t_{ij}$: Tiempo de viaje entre los nodos $i$ y $j$.

En cuanto a los costos:
- $c_{ij}$: Costo de transporte entre nodos $i$ y $j$.
- $f_t$: Costo de flete por kilómetro de cada vehículo $t$. 
- $v_t$: Costo por kilómetro del vehículo $t$. 

**TODO**: Agregar justificación.

https://www.motor.com.co/actualidad/Nueva-alza-en-la-gasolina-en-Colombia-aumento-95-pesos-en-febrero-2025-20250203-0002.html

https://cabify.com/co/tarifas/bogota#p-envios

https://www.kavak.com/mx/blog/camionetas-de-8-cilindros-potencia

Los costos que se asumen constantes durante todo el proceso son los siguientes. 

- Costo de mantenimiento ($M$): Debido a que para un camion el costo de mantenimiento cada 200 kilometros es de 119.000 pesos, realizando los calculos nos da $119,000/200km$ = $600 por km.
- Precio del combustible ($G$): Revisando las fuentes donde dice que en Colombia el precio de la gasolina esta en 16.184 pesos, y debido a que 1 camion consume 0.1 litros por km realizando los calculos nos da que un camion consume un precio de \$4.145 por km.
- Tarifa del servicio ($P$): \$600 por km este valor lo sacamos de las tarifas de envio de cabify.
- Consumo del combustible: 0,1 litros por km. Lo que significa 0,1 * G = 0,1*$15.000 = 1,500$
- Costo total por km = M+G+1.500

### 1.3. Variables de decisión y variables:

Las variables de decisión que se utilizarán en el modelo son las siguientes. 

1. $X_{i,j,t}$: Es una matriz binaria que indica si el vehículo $t$ vaija del nodo $i$ al nodo $j$. 1 si el vehículo $t$ recorre el arco de $i$ a $j$, de lo contrario $X_{i,j,t} = 0$.
2. $Q_{i,t}$: Cantidad de carga que transporta el vehículo $t$ cuando llega al nodo $i$. Es una variable continua que se mantiene en los límites de capacidad del vehículo. 
3. $P_{ti}$: Cantidad que no fue llevada del producto $i$ en el transporte $t$.

En cuanto a variables auxiliares que serán utiles para el problema se las plantean las siguientes dos.
1. $d_{i, t}$: Es la demanda atendida en el nodo $i$ por el vehículo $t$.
2. $u_{i, t}$: Orden de visita del nodo $i$ por el vehículo $t$ para eliminar subtours usando MTZ, lo cual garantiza la conectividad del recorrido.


### 1.4. Restricciones

1. La restricción de flujo, garantiza que si un vehiculo entra a un nodo también sale del mismo.
$$
\sum_{i \in N} X_{tij} = \sum_{j \in N} X_{tji}, \forall j \in C, \forall t
$$

2. La capacidad del vehículo, garantiza que la cantidad de productos transportados nunca supera la capacidad máxima de carga.
$$
\sum_{(i, j) \in N} d_j X_{tij} \leq Q_{tij}, \forall t
$$

3. Un cliente se atiende exactamente una vez, para simplificar el problema. 
$$
\sum_{i \in N} \sum_{t \in T} X_{tij} = 1, \forall j \in C
$$

4. Restricción sobre la cantidad transportada a un cliente no supere la demanda.
$$
Q_{tij} \leq d_j X_{tij}, \forall t \in T, \forall i \in N, \forall j \in C
$$

5. La carga transportada por un vehículo no supera la capacidad máxima.
$$
\sum_{j \in C} Q_{tij} \leq Q_{max} X_{tij}, \forall t \in T, \forall i \in N
$$

6. La cantidad entregada a un cliente $j$ sea igua a la demanda requerida $d_j$.
$$
\sum_{t \in T} \sum_{(i,j) \in N} Q_{tij} = d_j
$$

7. Conservación de flujo con las cantidades entregadas en cada uno de los nodos
$$
\sum_{j\in N} Q_{tij} - \sum_{j \in N} Q_{tji} = d_i, \forall t \in T, \forall i \in N \\ \{ 0 \}
$$

8. Cada vehículo no puede recorrer más distancia que la permitida por su autonomía máxima
$$
\sum_{(i,j) \in N} d_{ij} X_{tij} \leq a_{t} , \forall t \in T
$$


9. La cantidad total de productos enviados desde un centro de distribución no puede superar su capacidad
$$
\sum_{(i,j) \in N} Q_{tij} X_{tij} \leq S_i, \forall t \in T
$$

10. Para evitar ciclos en las rutas de los vehiculos definimos $u_i$ y aplicamos la restricción de Miller-Tucker-Zemlin (MTZ):
$$
u_i - u_j + |N| X_{tij} \leq |N| - 1, \forall (i,j) \in N
$$

11. Se agrega una restriccion para la penalizacion cunado no se cumple con la demanda requerida, donde M es un valor muy grande:

$$
(d_i - \sum_{(i,j) \in N} Q_{ijt}) \leq M*P_{ti}, \forall t \in V,\forall i \in N 
$$

## 2. Función Objetivo y Proceso de indagación

Para definir esta función objetivo, se llevó a cabo un análisis relacionado con la optimización de rutas y problemas logísticos. La formulación se diseñó con el objetivo de reducir al máximo los costos de transporte, los cuales están determinados por la distancia que se recorre y los costos operativos como tarifas de flete, mantenimiento y consumo de combustible.
El propósito es reducir al mínimo el costo total de transporte. Este costo se calcula teniendo en cuenta el valor por kilómetro recorrido, multiplicado por la distancia total que recorre cada vehículo en los trayectos seleccionados. Adicional, se le suma una penalizacion, cuando no se cumple con la demanda requerida. Por último, se agrego una restriccion para poder controlar el problema de los subtours.

La función objetivo está definida como:

$$
\min\left(\sum_{t \in T} \sum_{(i,j) \in N} C_{ij} \times X_{tij} + Penalizacion \right)
$$

## 3. Preprocesamiento de datos y análisis

En este punto se nos pide realizar la transformación de los datos ya que contamos con latitud y longitud pero se necesita el tiempo y las distancia. Una primera aproximación podría ser pensar en calcular la distancia euclidiana de estos puntos. Sin embargo, esto tiene un problema ya que obtenemos la distancia minima entre dos puntos en un plano, ignorando completamente calles y edificios, se podría pensar que los vehiculos atraviesan todo a su paso desde un centro de distribución hacía el punto de entrega, lo cual es incorrecto. Además de realizar la conversión a información que entienda el modelo se debe resolver este problema. Una solución bastante buena para este problema es utilizar el servicio de [`Distance Matrix API`](https://developers.google.com/maps/documentation/distance-matrix/overview) de Google Cloud, aprovechando que aún tenemos los créditos de arquisoft. 

In [4]:
try:
    import googlemaps
except Exception:
    print("googlemaps module not found")

En la siguiente celda se definen el código necesario para generar la matriz de distancias y la matriz de tiempo. Una vez se tiene esto se pueden calcular las matrices de costos facilmente unicamente multiplicando el valor que se definio por la matriz, gracias a que se esta trabajando con `numpy` el valor se multiplica de una vez por todas las entradas de la matriz.

In [5]:
# import numpy as np

# # Reemplaza con tu clave de API de Google Maps
# API_KEY = "AIzaSyDTD8eB5wVQyBDTBWUfA7aU9taCZ-T5nHg"
# gmaps = googlemaps.Client(key=API_KEY)

# ubicaciones = [
#     ("CD1", "Bodega Norte", 4.7110, -74.0721),
#     ("CD2", "Bodega Sur", 4.6050, -74.0835),
#     ("CD3", "Bodega Este", 4.6700, -74.0300),
#     ("C1", "Catalina", 4.6486, -74.0608),
#     ("C2", "Rodrigo", 4.7333, -74.0700),
#     ("C3", "Luis", 4.7000, -74.0200),
# ]

# # Crear un diccionario para mapear nombres a coordenadas
# coords = {loc[0]: (loc[2], loc[3]) for loc in ubicaciones}
# n = len(ubicaciones)  # Número total de ubicaciones

# # Inicializar matrices para distancias y tiempos
# distancias = np.zeros((n, n))
# tiempos = np.zeros((n, n))

# # Función para obtener distancia y tiempo entre dos ubicaciones
# def obtener_distancias_tiempos(origen, destino):
#     resultado = gmaps.distance_matrix(
#         origins=[origen],
#         destinations=[destino],
#         mode="driving",
#         units="metric"
#     )
#     distancia = resultado["rows"][0]["elements"][0]["distance"]["value"] / 1000  # km
#     tiempo = resultado["rows"][0]["elements"][0]["duration"]["value"] / 60  # minutos
#     return distancia, tiempo

# # Llenar la matriz de distancias y tiempos
# for i in range(n):
#     for j in range(n):
#         if i != j:  # No calcular la distancia de un nodo a sí mismo
#             dist, time = obtener_distancias_tiempos(coords[ubicaciones[i][0]], coords[ubicaciones[j][0]])
#             distancias[i, j] = dist
#             tiempos[i, j] = time

In [11]:
# # Imprimir matrices
# print("Matriz de Distancias (km):")
# print(distancias)

# print("\nMatriz de Tiempos (minutos):")
# print(tiempos)

Matriz de Distancias (km):
[[ 0.    14.504 15.445  9.23   2.727  7.516]
 [15.759  0.    18.667  7.623 18.211 17.42 ]
 [14.612 19.088  0.    10.077 17.339 14.046]
 [ 8.559  8.891  9.901  0.    12.34   9.802]
 [ 2.738 18.011 17.824 12.716  0.    10.247]
 [ 8.028 19.008 15.818 10.552  8.585  0.   ]]

Matriz de Tiempos (minutos):
[[ 0.         30.73333333 41.73333333 23.1         6.21666667 21.38333333]
 [30.25        0.         47.01666667 21.21666667 35.56666667 37.76666667]
 [38.21666667 47.76666667  0.         29.73333333 44.45       38.61666667]
 [24.7        25.05       31.13333333  0.         30.1        29.61666667]
 [ 6.06666667 35.85       48.06666667 31.6         0.         29.33333333]
 [23.53333333 43.25       46.48333333 32.98333333 27.05        0.        ]]


Además, dado que se esta buscando minimizar la función objetivo, no se puede utilizar tal cual la matriz anterior, se deben reemplazar los valores de 0 por un valor lo suficientemente grande como lo puede ser $999999$. 

In [12]:
# Matriz de Distancias (km)
distancias = [
    [999999, 14.504, 15.445, 9.23, 2.727, 7.516],
    [15.759, 999999, 18.667, 7.623, 18.211, 17.42],
    [14.612, 19.088, 999999, 10.077, 17.339, 14.046],
    [8.559, 8.891, 9.901, 999999, 12.34, 9.802],
    [2.738, 18.011, 17.824, 12.716, 999999, 10.247],
    [8.028, 19.008, 15.818, 10.552, 8.585, 999999]
]

# Matriz de Tiempos (minutos)
tiempos = [
    [999999, 30.73333333, 41.73333333, 23.1, 6.21666667, 21.38333333],
    [30.25, 999999, 47.01666667, 21.21666667, 35.56666667, 37.76666667],
    [38.21666667, 47.76666667, 999999, 29.73333333, 44.45, 38.61666667],
    [24.7, 25.05, 31.13333333, 999999, 30.1, 29.61666667],
    [6.06666667, 35.85, 48.06666667, 31.6, 999999, 29.33333333],
    [23.53333333, 43.25, 46.48333333, 32.98333333, 27.05, 999999]
]

## 4. Ejemplos ilustrativos

# Ejemplo 1: Caso básico con restricciones cumplidas

## Escenario:
- La bodega central es **Bodega Norte (CD1)**.
- Los clientes son **Catalina (C1)**, **Rodrigo (C2)** y **Luis (C3)**.
- La demanda de cada cliente es:
  - **C1:** 30 unidades.
  - **C2:** 20 unidades.
  - **C3:** 50 unidades.
- La capacidad máxima del vehículo es **100 unidades**.
- El costo por kilómetro es **3 UM/km**.
- El costo por minuto es **2 UM/min**.

### Matriz de Distancias (km):
|   | CD1  | C1   | C2   | C3   |
|---|-------|------|------|------|
| CD1 | 0     | 9.23 | 2.73 | 7.52 |
| C1  | 9.23  | 0    | 12.34| 9.80 |
| C2  | 2.73  | 12.34| 0    | 10.25|
| C3  | 7.52  | 9.80 | 10.25| 0    |

### Matriz de Tiempos (minutos):
|   | CD1  | C1   | C2   | C3   |
|---|-------|------|------|------|
| CD1 | 0     | 23.1 | 6.2 | 21.4 |
| C1  | 23.1  | 0    | 30.1| 29.6 |
| C2  | 6.2   | 30.1 | 0    | 29.3 |
| C3  | 21.4  | 29.6 | 29.3 | 0    |

### Ruta del Vehículo:
1. CD1 → C2 → C3 → C1 → CD1.

### Cálculo de Costos:
1. **Distancia Total Recorrida:** 2.73 + 10.25 + 9.80 + 9.23 = **32.01 km**.
2. **Tiempo Total:** 6.2 + 29.3 + 29.6 + 23.1 = **88.2 min**.
3. **Costos de Transporte:** 
   - Distancia: 32.01 × 3 = **96.03 UM**.
   - Tiempo: 88.2 × 2 = **176.4 UM**.
4. **Costo Total:** 96.03 + 176.4 = **272.43 UM**.

---

# Ejemplo 2: Penalización por unidades no entregadas

## Escenario:
- La bodega central es **Bodega Norte (CD1)**.
- Los clientes son **Catalina (C1)**, **Rodrigo (C2)** y **Luis (C3)**.
- La demanda de cada cliente es:
  - **C1:** 30 unidades.
  - **C2:** 40 unidades.
  - **C3:** 50 unidades.
- La capacidad máxima del vehículo es **100 unidades**.
- Penalización por cada unidad no entregada: **10 UM**.

### Distribución Propuesta:
1. CD1 → C3 → C2 → CD1.
2. **Cargas y Penalización:**
   - **C3:** 50 unidades entregadas (sin penalización).
   - **C2:** Solo 50 unidades pueden ser entregadas (demanda: 40, unidades entregadas: 50 → no penalización).
   - **C1:** No se entrega nada (demanda: 30, unidades no entregadas: 30 → penalización).

### Cálculo de Costos:
1. **Distancia Total Recorrida:** 7.52 + 10.25 + 2.73 = **20.5 km**.
2. **Costos de Transporte:** 20.5 km × 3 UM = **61.5 UM**.
3. **Penalización:** 30 unidades no entregadas × 10 UM = **300 UM**.
4. **Costo Total:** Transporte + Penalización = **61.5 + 300 = 361.5 UM**.

---

# Ejemplo 3: Restricción de capacidad no cumplida

## Escenario:
- La bodega central es **Bodega Norte (CD1)**.
- Los clientes son **Catalina (C1)**, **Rodrigo (C2)** y **Luis (C3)**.
- La demanda de cada cliente es:
  - **C1:** 40 unidades.
  - **C2:** 50 unidades.
  - **C3:** 20 unidades.
- La capacidad máxima del vehículo es **70 unidades**.

### Distribución Propuesta:
1. Primera ruta: CD1 → C1 → C2 → CD1 (70 unidades entregadas).
2. Segunda ruta: CD1 → C3 → CD1 (resto de la demanda).

### Restricciones no cumplidas:
- El vehículo no puede transportar más de 70 unidades, pero **la primera ruta excede la capacidad permitida**.

### Penalización:
1. El exceso de unidades debe ser transportado en otra ruta, incrementando el costo total.
2. **Costos adicionales estimados:**
   - Segunda ruta: CD1 → C3 → CD1.
   - Distancia total: 7.52 × 2 = **15.04 km**.

---

# Ejemplo 4: Restricción de capacidad no cumplida 

## Escenario:
- La bodega central es **Bodega Sur (CD2)**.
- Los clientes son **Catalina (C1)**, **Rodrigo (C2)** y **Luis (C3)**.
- La demanda de cada cliente es:
  - **C1:** 30 unidades.
  - **C2:** 40 unidades.
  - **C3:** 60 unidades.
- La capacidad máxima del vehículo es **50 unidades**.

### Matriz de Distancias (km):
|   | CD2  | C1   | C2   | C3   |
|---|-------|------|------|------|
| CD2 | 0     | 7.62 | 18.21 | 17.42 |
| C1  | 7.62  | 0    | 12.34 | 9.80  |
| C2  | 18.21 | 12.34| 0     | 10.25 |
| C3  | 17.42 | 9.80 | 10.25 | 0     |

### Matriz de Tiempos (minutos):
|   | CD2  | C1   | C2   | C3   |
|---|-------|------|------|------|
| CD2 | 0     | 21.2 | 35.6 | 37.7 |
| C1  | 21.2  | 0    | 30.1 | 29.6 |
| C2  | 35.6  | 30.1 | 0    | 29.3 |
| C3  | 37.7  | 29.6 | 29.3 | 0    |

### Distribución Propuesta:
1. Primera ruta: CD2 → C1 → C2 → CD2.
   - Total unidades entregadas: **50 unidades**.
   - Demanda: **C1 (30), C2 (40)**.
   - **Restricción incumplida:** No es posible cubrir toda la demanda de **C2** en esta ruta.

2. Segunda ruta: CD2 → C3 → CD2.
   - Total unidades entregadas: **50 unidades**.
   - Demanda: **C3 (60)**.
   - **Restricción incumplida:** Solo se entregan 50 unidades a **C3**, dejando una demanda insatisfecha de 10 unidades.

### Penalización:
1. **Unidades no entregadas:**
   - **C2:** 40 - 20 = 20 unidades no entregadas.
   - **C3:** 60 - 50 = 10 unidades no entregadas.
2. Penalización por unidades no entregadas: **10 UM por unidad**.
   - Total penalización: (20 + 10) × 10 = **300 UM**.

### Cálculo de Costos:
1. **Primera ruta (CD2 → C1 → C2 → CD2):**
   - Distancia total: 7.62 + 12.34 + 18.21 = **38.17 km**.
   - Tiempo total: 21.2 + 30.1 + 35.6 = **86.9 min**.
   - Costos:
     - Distancia: 38.17 × 3 = **114.51 UM**.
     - Tiempo: 86.9 × 2 = **173.8 UM**.
     - Subtotal: **288.31 UM**.
2. **Segunda ruta (CD2 → C3 → CD2):**
   - Distancia total: 17.42 × 2 = **34.84 km**.
   - Tiempo total: 37.7 × 2 = **75.4 min**.
   - Costos:
     - Distancia: 34.84 × 3 = **104.52 UM**.
     - Tiempo: 75.4 × 2 = **150.8 UM**.
     - Subtotal: **255.32 UM**.
3. **Costo Total:**
   - Transporte: 288.31 + 255.32 = **543.63 UM**.
   - Penalización: **300 UM**.
   - **Costo Total Final:** 543.63 + 300 = **843.63 UM**.


## 5. Justificación y discusión

**TODO**: Suposiciones que hay que hacer al modelo para simplificarlo. 